File: t/PACSFramework/CS/StorageServerManager/UI/AdministrativeSettings/Users/PasswordValidator.cs
// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Leadtools.Demos.StorageServer.DataTypes.Options;
using System.Text.RegularExpressions;

namespace Leadtools.Demos.StorageServer.UI.AdministrativeSettings.Users
{
   public static class PasswordValidator
   {
      public static bool Validate(string password, string confirmation, PasswordOptions options, out string validationMessage)
      {
         validationMessage = string.Empty;

         if(password.Length==0)
         {
            validationMessage = "Password cannot be empty.";
            return false;
         }

         if (password != confirmation)
         {
            validationMessage = "Password and confirm password do not match.";
            return false;
         }

         if (options.MinimumLength != 0)
         {
            if (password.Length < options.MinimumLength)
            {
               validationMessage = string.Format("Password must be at least {0} long.", options.MinimumLength);
               return false;
            }
         }

         if ((options.Complexity & Complexity.Lowercase) == Complexity.Lowercase)
         {
            if (!Regex.IsMatch(password, "[a-z]"))
            {
               validationMessage = string.Format("Password must must have at least one lower case letter.", options.MinimumLength);
               return false;
            }
         }

         if ((options.Complexity & Complexity.Uppercase) == Complexity.Uppercase)
         {
            if (!Regex.IsMatch(password, "[A-Z]"))
            {
               validationMessage = string.Format("Password must must have at least one upper case letter.", options.MinimumLength);
               return false;
            }
         }

         if ((options.Complexity & Complexity.Numbers) == Complexity.Numbers)
         {
            if (!Regex.IsMatch(password, "[0-9]"))
            {
               validationMessage = string.Format("Password must must have at least one number.", options.MinimumLength);
               return false;
            }
         }

         if ((options.Complexity & Complexity.Symbol) == Complexity.Symbol)
         {
            if (!Regex.IsMatch(password, "[@#$%^&+=]"))
            {
               validationMessage = string.Format("Password must must have at least one symbol.", options.MinimumLength);
               return false;
            }
         }

         return true;
      }
   }
}

File: t/PACSFramework/VB/StorageServerManager/UI/AdministrativeSettings/Users/PasswordValidator.vb
' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports Leadtools.Demos.StorageServer.DataTypes.Options
Imports System.Text.RegularExpressions

Namespace Leadtools.Demos.StorageServer.UI.AdministrativeSettings.Users
   Public Class PasswordValidator
      Private Sub New()
      End Sub
      Public Shared Function Validate(ByVal password As String, ByVal confirmation As String, ByVal options As PasswordOptions, <System.Runtime.InteropServices.Out()> ByRef validationMessage As String) As Boolean
         validationMessage = String.Empty

         If password.Length = 0 Then
            validationMessage = "Password cannot be empty."
            Return False
         End If

         If password <> confirmation Then
            validationMessage = "Password and confirm password do not match."
            Return False
         End If

         If options.MinimumLength <> 0 Then
            If password.Length < options.MinimumLength Then
               validationMessage = String.Format("Password must be at least {0} long.", options.MinimumLength)
               Return False
            End If
         End If

         If (options.Complexity And Complexity.Lowercase) = Complexity.Lowercase Then
            If (Not Regex.IsMatch(password, "[a-z]")) Then
               validationMessage = String.Format("Password must must have at least one lower case letter.", options.MinimumLength)
               Return False
            End If
         End If

         If (options.Complexity And Complexity.Uppercase) = Complexity.Uppercase Then
            If (Not Regex.IsMatch(password, "[A-Z]")) Then
               validationMessage = String.Format("Password must must have at least one upper case letter.", options.MinimumLength)
               Return False
            End If
         End If

         If (options.Complexity And Complexity.Numbers) = Complexity.Numbers Then
            If (Not Regex.IsMatch(password, "[0-9]")) Then
               validationMessage = String.Format("Password must must have at least one number.", options.MinimumLength)
               Return False
            End If
         End If

         If (options.Complexity And Complexity.Symbol) = Complexity.Symbol Then
            If (Not Regex.IsMatch(password, "[@#$%^&+=]")) Then
               validationMessage = String.Format("Password must must have at least one symbol.", options.MinimumLength)
               Return False
            End If
         End If

         Return True
      End Function
   End Class
End Namespace
